File: 2023/day-02/task2.mjs
import readline from 'node:readline';

function makeMax(globalRegex) {
  return string => Math.max(...Array.from(string.matchAll(globalRegex), m => parseInt(m[1], 10)));
}

const maxRed = makeMax(/(\d+?) red/g);
const maxGreen = makeMax(/(\d+?) green/g);
const maxBlue = makeMax(/(\d+?) blue/g);

let sum = 0;

for await (const line of readline.createInterface({ input: process.stdin })) {
  sum += maxRed(line) * maxGreen(line) * maxBlue(line);
}

console.log(sum);

File: 2023/day-02/task2.swift
var total = 0

while let line = readLine() {
  let splits = line.split(separator: " ")

  var maxRed = 0
  var maxGreen = 0
  var maxBlue = 0

  for i in 1..<splits.count {
    let chunk = splits[i]

    if (chunk.starts(with: "red")) {
      maxRed = max(Int(splits[i - 1])!, maxRed)
    } else if (chunk.starts(with: "green")) {
      maxGreen = max(Int(splits[i - 1])!, maxGreen)
    } else if (chunk.starts(with: "blue")) {
      maxBlue = max(Int(splits[i - 1])!, maxBlue)
    }
  }

  total += maxRed * maxGreen * maxBlue
}

print(total)
